### PR TITLE
feat: add keypoint/pose estimation support to RF-DETR

### PR DESCRIPTION
**Pull request by michaelmohamed**
Continuation of https://github.com/roboflow/rf-detr/pull/521 from develop 

Changes integrated into rfetr:develop latest 

 What does this PR do?

  Adds end-to-end keypoint (pose) estimation to RF-DETR. A new KeypointHead module predicts per-query (x, y, visibility) for each keypoint, integrated into the existing DETR detection pipeline
   with three-component training loss (L1 + BCE visibility + OKS) and COCO-format evaluation.

  Related Issue(s): Addresses community requests for pose estimation support in RF-DETR.

  Type of Change

  - New feature (non-breaking change that adds functionality)

  Summary

  - Add KeypointHead module with separate coordinate and visibility MLP heads that predict keypoints relative to detection bounding boxes
  - Add RFDETRPose model variants (Nano, Small, Medium, Large) with RFDETRPoseConfig and KeypointTrainConfig
  - Integrate three-component keypoint loss into SetCriterion: L1 coordinate loss (visible keypoints only), BCE visibility loss, and OKS (Object Keypoint Similarity) loss
  - Extend PostProcess to scale keypoint coordinates to pixel space and convert visibility logits to COCO format (0/2)
  - Add keypoint extraction to ConvertCoco and CocoDetection dataset classes with configurable num_keypoints
  - Add OKS metric accumulation and logging to COCOEvalCallback (val table + TensorBoard)
  - Add pose estimation documentation page and API reference

  Architecture

  Transformer Query Features → KeypointHead
    ├── coord_head (MLP) → sigmoid → [0,1] coords → box-relative offset
    └── visibility_head (MLP) → logits → sigmoid at inference
  → [B, num_queries, num_keypoints, 3]

  Testing

  - I have tested this change locally
  - I have added/updated tests for this change

  Test details:

  pytest tests/test_keypoint_head.py -v   # tests, all passing

  Test coverage includes:
  - KeypointHead forward pass, reference-box mode, multi-layer, custom keypoint counts
  - COCO constants validation (names, skeleton, sigmas, flip pairs)
  - Config classes (RFDETRPoseConfig, KeypointTrainConfig)
  - Loss function integration (loss_keypoints in SetCriterion)
  - Post-processing with/without keypoints
  - Inference output structure, visibility sigmoid, coordinate scaling
  - Gradient flow through keypoint head

  Checklist

  - My code follows the style guidelines of this project
  - I have performed a self-review of my own code
  - I have commented my code where necessary, particularly in hard-to-understand areas
  - My changes generate no new warnings or errors
  - I have updated the documentation accordingly (if applicable)

### CODE REVIEW

**Comment by Borda** [comment]
> Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/roboflow/rf-detr?pullRequest=868) before we can accept your contribution.

@michaelmohamed ^^ 🦝

**Comment by Borda** [comment]
@michaelmohamed pls, please stop pushing commits until you sign the CLA. We cannot merge your work...

**Comment by michaelmohamed** [comment]
@Borda 
image

**Comment by Borda** [comment]
> @Borda image

Perfect, thank you! :)
now passing to @isaacrob to guide arch staff 🦝

**Comment by isaacrob** [comment]
Hi! As outlined in the contribution guide, an architecture change requires demonstration on common benchmark that it outperforms comparable models in order to be merged. Do you have any such benchmarks?

As an aside, we will be releasing a keypoint model soon :)

**Comment by isaacrob** [comment]
From my perspective the most directly comparable model is YOLO26

**Comment by michaelmohamed** [comment]
@isaacrob i was hoping that someone could help me with benchmarking. 

I tried to mimic the implementation of yolov11 and yolov26 in this setup. 

Since you are working on an implementation that will be released soon, and you'll benchmark it, closing this PR. 

Thanks.

File: src/rfdetr/models/criterion.py [hunk 1]
```diff
@@ -446,6 +446,138 @@ def loss_masks(self, outputs, targets, indices, num_boxes):
...
+        else:
+            coord_loss = src_coords.sum() * 0
+
+        # BCE loss for visibility
+        target_vis_binary = (target_vis > 0).float()
+        vis_loss = F.binary_cross_entropy_with_logits(src_vis_logits, target_vis_binary, reduction="mean")
+
+        # OKS loss (Object Keypoint Similarity)
+        oks_loss = self._compute_oks_loss(src_coords, target_coords, valid_mask, targets, indices)
+
```
[the remark this answers is not in the document]
Comment (michaelmohamed) [reply src/rfdetr/models/criterion.py:514]: All three keypoint losses now use sum / num_boxes normalization, consistent with loss_bbox and loss_giou in the same criterion.

File: src/rfdetr/datasets/coco.py [hunk 2]
```diff
@@ -249,11 +264,59 @@ def __call__(self, image: Image.Image, target: Dict[str, Any]) -> Tuple[Image.Im
...
+            is empty.
+        """
+        if len(anno) == 0:
+            return torch.zeros((0, self.num_keypoints, 3), dtype=torch.float32)
+
+        all_keypoints = []
+        for obj in anno:
+            if "keypoints" in obj and len(obj["keypoints"]) > 0:
+                kpts = obj["keypoints"]
+                # COCO format: [x1, y1, v1, x2, y2, v2, ...]
```
[the remark this answers is not in the document]
Comment (michaelmohamed) [reply src/rfdetr/datasets/coco.py:301]: Added validation that len(kpts) % 3 == 0 before the reshape, with a clear error message including the annotation ID

**Comment by isaacrob** [comment]
@michaelmohamed i would encourage you to benchmark your work regardless :) very valuable exercise. And it's always possible it beats ours!

What do you need help with?

**Comment by michaelmohamed** [comment]
> @michaelmohamed i would encourage you to benchmark your work regardless :) very valuable exercise. And it's always possible it beats ours!
> 
> What do you need help with?

Is there a standard dataset and parameters that you all will leverage for testing? @isaacrob

**Comment by isaacrob** [comment]
I'm benchmarking on COCO pose dataset, seems like the most standard. what do you mean by parameters? the goal is to have a model that is Pareto optimal over existing approaches, ie either faster or more accurate (or both!) vs prior art

**Comment by michaelmohamed** [comment]
Got it. Are there any hyper parameters that we set to ensure parity across tests?

Do you have a timeline around release of your keypoint/pose implementation?

Thank you!

@isaacrob

**Comment by isaacrob** [comment]
I mean not really :) lots of people use different settings for training their COCO models. IMO the only thing that matters is final performance ie latency vs accuracy tradeoff

I don't want to commit to a timeframe on release. But making steady progress.

**Comment by michaelmohamed** [comment]
@isaacrob 

Got it. Not to hold you to a specific date, but perhaps the quarter that you are thinking. I know that there a lot of folks tracking this.

Do you have a branch that you are working off of that you can share for reference? I'm sure there are others that would like to support where possible. Thanks!

Regards, 

Michael

**Comment by isaacrob** [comment]
Aiming for this quarter but we'll see what happens! Very close to SOTA, won't release till we've won. Already have SOTA on some latency ranges but not yet all of them.  

I'm working in a different codebase for now. Eventually we'll unify so we get the benefits of @Borda 's work as well

**Comment by michaelmohamed** [comment]
Sounds awesome. Appreciate the heads up. Please don't hesitate to reach out if you need help testing. Thanks again!

**Comment by C-Moric** [comment]
@isaacrob Any updates on the current state of the keypoint head?

**Comment by Borda** [comment]
Let's revise the contribution policy and accept these as experimental or community with clear separation in docs...

**Comment by xingshou** [comment]
When is the keypoint/pose estimation feature gonna be added to the official version?
